netket/hilbert/homogeneous.py

```diff
@@ -182,31 +179,37 @@ def all_states(self) -> np.ndarray:
             A (n_states x size) batch of states. this corresponds
             to the pre-allocated array if it was passed.
         """
+        if not self.is_indexable:  # includes call to _setup
+            raise RuntimeError("The hilbert space is too large to be indexed.")
 
         return self._hilbert_index.all_states()
 
     @property
     def _hilbert_index(self) -> HilbertIndex:
-        """
-        Returns the `HilbertIndex` object, which is a numba jitclass used to convert
-        integers to states and vice-versa.
-        """
-        if self.__hilbert_index is None:
-            if not self.is_indexable:
-                raise RuntimeError("The hilbert space is too large to be indexed.")
-
+        if self._hilbert_index_ is None:
+            # the unconstrained index
+            index = UniformTensorProductHilbertIndex(self._local_states, self.size)
             if self.constrained:
```

Comment:
```suggestion
            if self.constrained:
            # If we have a constraints, we tentatively construct a specialised Hilbert index for that particular constraint.
            # If this specialised indexer object exists, we check whether it is more efficient than the generic
            # ConstrainedHilbertIndex one. If it is more efficient, we use it, otherwise we keep the generic one.
            
```